include/ginkgo/core/base/batch_multi_vector.hpp

```diff
@@ -0,0 +1,813 @@
+ *                (not including the implied Executor as the first argument)
+ *
+ * @param vals  values used to initialize the batch vector
+ * @param exec  Executor associated to the vector
+ * @param create_args  additional arguments passed to Matrix::create, not
+ *                     including the Executor, which is passed as the first
+ *                     argument
+ *
+ * @ingroup BatchMultiVector
+ * @ingroup mat_formats
+ */
+template <typename Matrix, typename... TArgs>
+std::unique_ptr<Matrix> batch_initialize(
+    std::initializer_list<std::initializer_list<typename Matrix::value_type>>
+        vals,
+    std::shared_ptr<const Executor> exec, TArgs&&... create_args)
+{
+    using batch_multi_vector = BatchMultiVector<typename Matrix::value_type>;
+    size_type num_batch_entries = vals.size();
+    GKO_ASSERT(num_batch_entries > 0);
+    auto vals_begin = begin(vals);
+    size_type common_num_rows = vals_begin->size();
+    auto common_size = dim<2>(common_num_rows, 1);
+    for (auto& val : vals) {
+        GKO_ASSERT_EQ(common_num_rows, val.size());
+    }
+    auto b_size = batch_dim<2>(num_batch_entries, common_size);
+    auto tmp = batch_multi_vector::create(exec->get_master(), b_size);
+    size_type batch = 0;
+    for (const auto& b : vals) {
+        size_type idx = 0;
+        for (const auto& elem : b) {
+            tmp->at(batch, idx) = elem;
+            ++idx;
+        }
+        ++batch;
+    }
+    auto mtx = Matrix::create(exec, std::forward<TArgs>(create_args)...);
+    tmp->move_to(mtx);
+    return mtx;
+}
+
+
+/**
+ * Creates and initializes a batch of matrices.
+ *
+ * This function first creates a temporary Dense matrix, fills it with passed in
+ * values, and then converts the vector to the requested type.
+ *
+ * @tparam Matrix  matrix type to initialize
+ *                 (Dense has to implement the ConvertibleTo<Matrix> interface)
+ * @tparam TArgs  argument types for Matrix::create method
+ *                (not including the implied Executor as the first argument)
+ *
+ * @param vals  values used to initialize the vector
+ * @param exec  Executor associated to the vector
+ * @param create_args  additional arguments passed to Matrix::create, not
+ *                     including the Executor, which is passed as the first
+ *                     argument
+ *
+ * @ingroup BatchMultiVector
+ * @ingroup mat_formats
+ */
+template <typename Matrix, typename... TArgs>
+std::unique_ptr<Matrix> batch_initialize(
+    std::initializer_list<std::initializer_list<
+        std::initializer_list<typename Matrix::value_type>>>
+        vals,
+    std::shared_ptr<const Executor> exec, TArgs&&... create_args)
+{
+    using batch_multi_vector = BatchMultiVector<typename Matrix::value_type>;
+    size_type num_batch_entries = vals.size();
+    GKO_ASSERT(num_batch_entries > 0);
+    auto vals_begin = begin(vals);
+    size_type common_num_rows = vals_begin->size();
```

Comment:
Alternative without assertions:
```suggestion
    size_type common_num_rows = vals_begin ? vals_begin->size() : 0;
```
This would allow to create 0x0 batches.